server/cluster/unsafe_recovery_controller.go

```diff
@@ -125,12 +128,216 @@ func (u *unsafeRecoveryController) HandleStoreHeartbeat(heartbeat *pdpb.StoreHea
+		// to create new regions for each of the gap, besides, merge the all the scattered
+		// ranges that are connected by filling up these gaps.
+		overlapRanges := make([]*metapb.Region, 0)
+		recoveredRanges.DescendLessOrEqual(regionItem{region}, func(item btree.Item) bool {
+			if bytes.Compare(item.(regionItem).region.StartKey, region.StartKey) < 0 {
+				// Only adds the first region that has smaller start key to the
+				// *potential* overlap regions list.
+				overlapRanges = append(overlapRanges, item.(regionItem).region)
+			}
+			return false
+		})
+
+		recoveredRanges.AscendGreaterOrEqual(regionItem{region}, func(item btree.Item) bool {
+			if len(region.EndKey) != 0 && bytes.Compare(item.(regionItem).region.StartKey, region.EndKey) > 0 {
+				return false
+			}
+			overlapRanges = append(overlapRanges, item.(regionItem).region)
+			return true
+		})
+		lastEnd := region.StartKey
+		reachedTheEnd := false
+		var creates []*metapb.Region
+		for _, overlapRegion := range overlapRanges {
+			if bytes.Compare(lastEnd, overlapRegion.StartKey) < 0 {
+				newRegion := proto.Clone(region).(*metapb.Region)
+				newRegion.StartKey = lastEnd
+				newRegion.EndKey = overlapRegion.StartKey
+				newRegion.Id, _ = u.cluster.AllocID()
+				creates = append(creates, newRegion)
+				recoveredRanges.ReplaceOrInsert(regionItem{newRegion})
+				if len(overlapRegion.EndKey) == 0 {
+					reachedTheEnd = true
+					break
+				}
+				lastEnd = overlapRegion.EndKey
+			} else if len(overlapRegion.EndKey) == 0 {
+				reachedTheEnd = true
+				break
+			} else if bytes.Compare(overlapRegion.EndKey, lastEnd) > 0 {
+				lastEnd = overlapRegion.EndKey
+			}
+		}
+		if !reachedTheEnd && (bytes.Compare(lastEnd, region.EndKey) < 0 || len(region.EndKey) == 0) {
+			newRegion := proto.Clone(region).(*metapb.Region)
+			newRegion.Id, _ = u.cluster.AllocID()
+			newRegion.StartKey = lastEnd
+			newRegion.EndKey = region.EndKey
+			creates = append(creates, newRegion)
+			recoveredRanges.ReplaceOrInsert(regionItem{newRegion})
+		}
+		if len(creates) != 0 {
+			storeRecoveryPlan, exists := u.storeRecoveryPlans[storeId]
+			if !exists {
+				u.storeRecoveryPlans[storeId] = &pdpb.RecoveryPlan{}
+				storeRecoveryPlan = u.storeRecoveryPlans[storeId]
+			}
+			storeRecoveryPlan.Creates = append(storeRecoveryPlan.Creates, creates...)
+		}
+		if _, healthy := healthyRegions[region.Id]; healthy {
+			if updated {
+				storeRecoveryPlan, exists := u.storeRecoveryPlans[storeId]
+				if !exists {
+					u.storeRecoveryPlans[storeId] = &pdpb.RecoveryPlan{}
+					storeRecoveryPlan = u.storeRecoveryPlans[storeId]
+				}
+				storeRecoveryPlan.Updates = append(storeRecoveryPlan.Updates, region)
+			}
+		} else {
+			storeRecoveryPlan, exists := u.storeRecoveryPlans[storeId]
+			if !exists {
+				u.storeRecoveryPlans[storeId] = &pdpb.RecoveryPlan{}
+				storeRecoveryPlan = u.storeRecoveryPlans[storeId]
+			}
+			storeRecoveryPlan.Deletes = append(storeRecoveryPlan.Deletes, region.Id)
```

Comment:
why delete unhealthy regions

Reply:
This is just for simplifying the implementation, that, by doing this, we don't have to add extra code to decide which piece of the data stays at the original region, which piece of the data is going to be used as the base of newly created region, imagine a certain peer (or a range) is cut into several different pieces.

Reply:
but for the updated region, the process of it may be different from the ones newly created.

Reply:
Updated, now it reuses region ID if possible